src/s2/base/spinlock.h

```diff
@@ -16,39 +16,43 @@
 #ifndef S2_BASE_SPINLOCK_H_
 #define S2_BASE_SPINLOCK_H_
 
-#include <atomic>
+#include <absl/base/thread_annotations.h>
 
+#include <atomic>
 
-class SpinLock {
+class ABSL_LOCKABLE SpinLock {
  public:
   SpinLock() = default;
   ~SpinLock() = default;
   SpinLock(SpinLock const&) = delete;
   SpinLock& operator=(SpinLock const&) = delete;
 
-  inline void Lock() {
+  inline void Lock() ABSL_EXCLUSIVE_LOCK_FUNCTION() {
     while (locked_.exchange(true, std::memory_order_acquire)) {
       // Spin.
       continue;
     }
   }
 
-  inline void Unlock() {
+  inline void Unlock() ABSL_UNLOCK_FUNCTION() {
     locked_.store(false, std::memory_order_release);
   }
 
-  inline bool IsHeld() const {
+  inline bool IsHeld() const ABSL_ASSERT_EXCLUSIVE_LOCK() {
```

Comment:
I didn't look at this closely enough.

`ABSL_ASSERT_EXCLUSIVE_LOCK` is for a function that `assert()`s / terminates.

https://clang.llvm.org/docs/ThreadSafetyAnalysis.html#assert-capability-and-assert-shared-capability

Can you remove `ABSL_ASSERT_EXCLUSIVE_LOCK()` for now or make an `AssertHeld()` function and use that?

https://github.com/abseil/abseil-cpp/blob/9ac131cf7da4de8b19b4c956a0c13edba74d92d9/absl/base/internal/spinlock.h#L129

Reply:
About assertheld I'm not sure, because usecase in s2 different:
`ABSL_DCHECK(cells_lock_.IsHeld());`